src/main/java/de/tum/in/www1/artemis/service/scheduled/ProgrammingExerciseScheduleService.java

```diff
@@ -241,6 +262,30 @@ public Runnable lockAllStudentRepositories(ProgrammingExercise exercise) {
         return lockStudentRepositories(exercise, participation -> true);
     }
 
+    /**
+     * Returns a runnable, that, once executed, will update all results for the given exercise.
+     *
+     * This might be needed for an exercise that has test cases marked with
+     * {@link de.tum.in.www1.artemis.domain.enumeration.Visibility#AFTER_DUE_DATE}.
+     *
+     * Those test cases might already have been run in the continuous integration
+     * service and their feedbacks are therefore stored in the database.
+     * However, they are not included in the student score before the due date has passed.
+     * Updating the student score includes the feedbacks of those test cases into
+     * the result without having to trigger a new continuous integration job.
+     *
+     * @param exercise the exercise for which the results should be updated
+     * @return a Runnable that will update all results for the given exercise
+     */
+    @NotNull
+    public Runnable updateAllStudentScores(ProgrammingExercise exercise) {
```

Comment:
Thanks for the explanation above. Can we make sure this is only invoked for exercises which have used the feature that tests are visible only after the due date? (Many programming exercises will not use this feature and I would like to avoid expensive calculations then which might even lead to race conditions)

Also I am still not sure if this is really needed, in case the builds are executed afterwards anyway. In the previous configuration before this PR, we also had tests that have only been used afterwards. However, all builds have been executed anyway on the build and test run after due date. This has actually created a new result.

I assume this will still be the case, so why should we recalculate the old results? (which will be outdated anyway, right?)

Reply:
I changed the behaviour to the following:
- if an exercise has a `buildAndTestAfterDueDate`: no score updates are scheduled independent of test case visibilities, the rebuild takes care of any needed score updates.
- if an exercise has no tests marked as `AFTER_DUE_DATE`: no score updates are scheduled, as they are not needed.
- if an exercise has tests marked as `AFTER_DUE_DATE` but no `buildAndTestAfterDueDate`: a score update is scheduled, as the tests have to become visible somehow.

That way no race conditions on changed scores should be possible and recalculations are avoided where possible.